### DISCUSSION TITLE
'[SUB-90011] Subject does not match consumer configuration filter.'

### DISCUSSION THREAD

**Discussion by Exagram** [nats-io/nats.net.v1#568] [Q&A] [answered]
## Defect

Make sure that these boxes are checked before submitting your issue -- thank you!

 - [X] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### Versions of `NATS.Client` and `nats-server`:

- nats-0.0.28-windows-amd64
- nats-server-v2.7.2-windows-amd64

#### OS/Container environment:

- Windows 10 OS

#### Steps or code to reproduce the issue:

I have registered a stream:

```powershell
.\nats.exe str add STREAM --subjects "STREAM.>" --ack --retention "limits" --replicas 1 --discard old --dupe-window="2m0s" --max-msgs=-1 --max-bytes=-1 --max-age=1y --max-msg-size=-1 --max-consumers=-1 --storage file --max-msgs-per-subject=-1 --deny-purge=false --deny-delete=false --allow-rollup-hdrs=false
```

I have registered a consumer:

```powershell
.\nats.exe consumer add STREAM CONSUMER1 --pull --filter "STREAM.>" --deliver last --ack explicit --replay instant --sample 100 --max-pending 0 --max-deliver=-1
```

I wrote C# code for the pull-based consumer scenario:

```
var context = connection.CreateJetStreamContext();
var options = PullSubscribeOptions.Builder()
    .WithStream("STREAM")
    .WithDurable("CONSUMER1")
    .Build();
var subscription = context.PullSubscribe("STREAM.entity1.*.failed", options);
var msgs = subscription.Fetch(1, 1000);
...
```

#### Expected result:

Since `STREAM.entity1.*.failed` is a subset of `STREAM.>` I expect the `PullSubscribe` to throw zero errors.

#### Actual result:

`PullSubscribe` throws an exception: `NATS.Client.NATSJetStreamClientException: '[SUB-90011] Subject does not match consumer configuration filter.'`

Note: There is no documentation for `PullSubscribe(subject)` and especially its first parameter "subject" in https://docs.nats.io/nats-concepts/jetstream/js_walkthrough. Instead I see `nats consumer next {STREAM} {CONSUMER}`. Where's the subject in this command?

**Comment by scottf** [comment, marked as answer]
@Exagram So while it's true that `STREAM.entity1.*.failed` is a subset of `STREAM.>` , we require an exact match, which means the code is behaving as designed. The server team is currently working on enhancements to how subjects are consumed providing better flexibility on filter subjects., so this will eventually trickle down to the clients. For now, your consumer filter subject must match exactly the subscribe subject, so you must use  `STREAM.entity1.*.failed` in both places.

You are certainly not the first to notice this, we probably get questions like this daily, so I'm moving this issue to our discussion area. I agree, we need to do better with our documentation, it's an ongoing effort.

**Reply by Walluce** [reply]
Would it be at all possible to get an update on this?

**Reply by scottf** [reply]
I think I confused the issue. The flexibility I was thinking about does not apply here. You simply must make the subscription subject match the consumer subject. Since your consumer is pre-made, you might consider using BindTo for the options and null as the subscribe subject
```
PullSubscribeOptions pso = PullSubscribeOptions.BindTo("STREAM", "CONSUMER1");
js.PullSubscribe(null, pso)
```
